router/gossip.go

```diff
@@ -0,0 +1,137 @@
+	Gossip() []byte
+	// merge in state and return "everything new I've just learnt",
+	// or nil if nothing in the received message was new
+	OnGossip(buf []byte) []byte
+}
+
+type GossipChannel struct {
+	localPeer *LocalPeer
+	name      string
+	hash      uint32
+	gossiper  Gossiper
+}
+
+func (router *Router) NewGossip(channelName string, g Gossiper) Gossip {
+	h := hash(channelName)
+	channel := &GossipChannel{router.Ourself, channelName, h, g}
+	router.GossipChannels[h] = channel
+	return channel
+}
+
+func (router *Router) SendAllGossip() {
+	for _, c := range router.GossipChannels {
+		c.GossipMsg(c.gossiper.Gossip())
+	}
+}
+
+func (router *Router) SendGossip(channelName string, msg []byte) {
+	channelHash := hash(channelName)
+	if channel, found := router.GossipChannels[channelHash]; !found {
+		log.Println("[gossip] attempt to send on unknown channel:", channelName)
+	} else {
+		channel.GossipMsg(msg)
+	}
+}
+
+func (c *GossipChannel) GossipMsg(buf []byte) {
+	c.localPeer.ForEachConnection(func(_ PeerName, conn Connection) {
+		if conn.Established() {
+			peerName := c.localPeer.Name.Bin()
+			nameLenByte := []byte{byte(len(peerName))}
+			msg := Concat([]byte{ProtocolGossip}, uint32slice(c.hash), nameLenByte, peerName, buf)
+			conn.(ConnectionSender).SendTCP(msg)
+		}
+	})
+}
+
+func handleGossip(conn *LocalConnection, msg []byte, onok func(channel *GossipChannel, srcName PeerName, origMsg, payload []byte)) {
+	if len(msg) < 10 {
+		conn.log("[gossip] received invalid message of length:", len(msg))
+		return
+	}
+	channelHash, payload := decodeGossipChannel(msg[1:])
+	if channel, found := conn.Router.GossipChannels[channelHash]; !found {
+		conn.log("[gossip] received unknown channel:", channelHash)
+	} else {
+		srcName, payload := decodePeerName(payload)
+		onok(channel, srcName, msg, payload)
+	}
+}
+
+func deliverGossipUnicast(channel *GossipChannel, srcName PeerName, origMsg, payload []byte) {
+	destName, msg := decodePeerName(payload)
+	if channel.localPeer.Name == destName {
+		channel.gossiper.OnGossipUnicast(srcName, msg)
+	} else {
+		channel.RelayGossipTo(destName, origMsg)
+	}
+}
+
+func deliverGossipBroadcast(channel *GossipChannel, srcName PeerName, origMsg, payload []byte) {
+	channel.gossiper.OnGossipBroadcast(payload)
+	channel.RelayGossipBroadcast(srcName, origMsg)
+}
+
+func deliverGossip(channel *GossipChannel, srcName PeerName, origMsg, payload []byte) {
+	if newBuf := channel.gossiper.OnGossip(payload); newBuf != nil {
+		channel.GossipMsg(newBuf)
+	}
+}
+
+func (c *GossipChannel) GossipUnicast(dstPeerName PeerName, buf []byte) error {
+	srcPeerByte := c.localPeer.Name.Bin()
+	nameLenByte := []byte{byte(len(srcPeerByte))}
+	dstPeerByte := dstPeerName.Bin()
+	dstNameLenByte := []byte{byte(len(dstPeerByte))}
+	msg := Concat([]byte{ProtocolGossipUnicast}, uint32slice(c.hash), nameLenByte, srcPeerByte, dstNameLenByte, dstPeerByte, buf)
+	return c.RelayGossipTo(dstPeerName, msg)
+}
```

Comment:
It takes 17 bytes to encode the basic gossip header, as opposed to 12 with this implementation.  We do use non-gob encodings, e.g. [here](https://github.com/zettio/weave/blob/1b229fcc30c45fb4c5e06c6a85d0e8e914feed6a/router/crypto.go#L133) and [here](https://github.com/zettio/weave/blob/1b3d65476c23de7f6242f32b924cd142bdb0110d/router/connection.go#L180).

That said, it would make a better argument if this code was similar to either of those, possibly using encoding/binary.